Duckdb query engine

Mittmich opened the pull request:
This PR implements [this ticket](https://trello.com/c/dqBhx8vJ/91-implement-select-query-flow-using-duckdb). It suggests the class structure of the query engine (can be found in docs\query_engine_interface.md) and provides and implementation using duckdb. A good place to start the review is the usage documentation that can be found here: docs\query_engine_usage.md. Looking forward to your comment!

dmitrymyl commented:
Very much thanks @Mittmich for the pull request! I am looking at it now and will split my review in two parts: concepts and code.

dmitrymyl commented:
# Concepts

## Overview

I really like the proposed architecture! I haven't looked at the code yet, but it seems sensible and self-explanatory based on the UML diagram. I value that queries are flexible and composable. Docs are well-written and it is quite clear how to use the engine. The architecture seems alright, I have some suggestions about naming and some small questions as well.

## Naming suggestions
1. `query = BasicQuery([...]); result = query.query(dataset)`. `query.query` sounds redundant. Maybe we should just implement `__call__` to make it `query(dataset)`? Although, this stuff is a bit advanced python feature, less experienced users might not be familiar with it and might find it confusing.
2. `result.compute()` sounds better and in line with dask than `result.load_result()`.
3. `Snipper`, `Aggregation`, `Transformation` &mdash; imo verbs would be shorter to type and they are also in line with SQL syntax (thus helping with adaption to our query engine). So I suggest to rename all `QueryStep` into verbs like `Snip`, `Aggregate`, `Transform`.
4. I feel reluctant about `Snip` naming. IMO this operation represents selection of multi-way contacts in one or many genomic regions, not necessarily followed by pileups. So I would suggest to rename it to `Select`. It might get confused with the similar SQL operator, but the same word is used in [`bioframe.select`](https://bioframe.readthedocs.io/en/latest/guide-intervalops.html#selecting-slicing). I couldn't find any similar name in other packages (bedtools, granges, samtools), they use words "intersect", "overlap". Other potential names could be "subset" or "extract". Some other reasons to not use `Snip`:
    1. Currently, I am not sure about broad applicability of pileups in multi-way contacts analysis, since they are averaging and hinder diversity of the data (I will send my last lab presentation in Teams). I don't think they will be a frequent type of analysis, since there are way more other available and potential analytical methods.
    2. Snipping doesn't seem to be a broadly acceptable term in the Hi-C field. It is used in cooltools code, but not in two other popular python packages (FAN-C and HiCExplorer -- they use "aggregation" and "average"). The first page of google search for `Hi-C "snipping"` contains links to only cooltools manual and papers by Mirny lab.
5. In the docs, `Filter along the rows` &mdash; I have always found this terminology confusing in pandas. Does it subset rows or subset columns? I suspect the former. It could also be rewritten from the perspective of contacts and individual fragments, but this is not crucial ATM.
6. I was thinking about some generic object to be named as `BasicQuery`. But it is a central public object now, not basic. So we should rename it in the future for the public release (e.g., `Query`, `QueryEngine`, ...). But not now!


## Questions
7. What are `get_position_coordinates()` of `gDataSchema`? It was not described in the UML diagram (maybe it will be clear after I check the code).
8. Mermaid diagram: should all arrows be in reverse? It's the case with [UML diagrams](https://www.visual-paradigm.com/guide/uml-unified-modeling-language/uml-class-diagram-tutorial/), but maybe syntax is different here. Anyway, it doesn't matter for the project per se.
9. It was a bit confusing that `QueryResult` is an output of the `BasicQuery` and implements the `gDataProtocol` at the same time. Initially, I was thinking that the `QueryResult` is an actual materialization, like a 2D pileup matrix. But it is actually a lazy evaluated result, not the data. Then it makes sense to say it implements the `gDataProtocol`.
10. However, I am still not sure that the `QueryResult` should implement the `gDataProtocol`. It makes sense since the `BasicQuery` can accept the `QueryResult`. But what if the `QueryResult` actually relates to smth not being `Pixels` or `Contacts`, like 2D pileup matrix? What sense is there for it to follow the `gDataProtocol`? As an alternative, `QueryResult` might or might not implement the `gDataProtocol` to accomodate for this ambiguity.
11. `BasicQuery.query()` builds the query plan for postponed execution. The query plan is specified in the class instance and is applied to a dataset. This implies that query steps are independent of the dataset. Can it be otherwise: that features of the dataset (shape, contact orders etc) dynamically dictate the specific choice of procedures (in a way that user doesn't know that beforehand and cannot write another query plan)?

dmitrymyl commented on spoc/contacts.py:
Code changes are due to adding duckdb backend support to Contacts and Pixels. Does it need to be though? Does it represent an intermediate object of a query or a final result of a query (yet to be evaluated)? Can there be an alternative class for that?

Mittmich replied:
Yes, so my thoughts are the following: In the long run, get rid of pandas/dask everywhere except for specific manifestations of data. To achieve this, we need to switch all classes that pass around data to handling `duckdb.DuckDBPyRelation` objects. This represents the result of a duckdb query and - importantly -, are not executed yet. They are lazy, but implement looking at the data's schema via the `.columns` method. To execute them, one can either write them to file via `.to_parquet`, or call `.df()`. 

So an ideal flow would be:
- `FileManager` loads data as `duckdb.DuckDBPyRelation` objects and instantiates Contacts and Pixels
- Contacts/Pixels feed the data into the query engine that anyway operates on  `duckdb.DuckDBPyRelation`
- If Contacts/Pixels need to be persisted, FileManager takes a  `duckdb.DuckDBPyRelation` and writes it to parquet.

dmitrymyl replied:
Okay, that sounds good. I can't really imagine frequent cases when a user would need to have direct access to dataframes as python objects (again, similar to .cool) -- except for debugging or diagnostics. So this should work!

dmitrymyl commented on spoc/io.py:
```diff
@@ -243,10 +243,7 @@ def load_pixels(
-            df = self._parquet_reader_func(pixel_path)
-        else:
-            df = pixel_path
+        df = self._parquet_reader_func(pixel_path)
```
Does this assume that a dataframe will always be loaded into RAM? Because before that we could use just a filename and load when needed. So, how does this modification affect use cases?

Mittmich replied:
No, it does not! The removal of fileanme makes the first step of having everything depend on  `duckdb.DuckDBPyRelation`, which is lazy. The filename was just a hack to achieve that. With  `duckdb.DuckDBPyRelation`, we can have the laziness, but also get information about the schema of the data.

dmitrymyl replied:
But how would user instantiate Pixels/Contacts? And also how to use duckdb in this case?

Mittmich replied:
Ah yes, I see your point, sorry, I thought I added `duckdb.read_parquet` to the reader functions. Will do asap!

Mittmich replied:
Added it :)

dmitrymyl commented on spoc/models/dataframe_models.py:
```diff
@@ -29,6 +31,73 @@
+        """Returns the position fields as a dictionary
+        of framgent index to the respective fields"""
+
+    def get_contact_order(self) -> int:
```
This is applicable to single tables of one contact order. How does it scale to a whole dataset with many contact orders?

In other words, if I would like to display all multi-way contacts within a selected genomic region, how will the corresponding procedure be implemented? Will it be a loop over all tables with different contact orders?

Mittmich replied:
Indeed, that would be the implementation. We can think of having somting like a compound object that contains files of multiple orders, but under the hood, I think that compound oject would get the position fields of each of the orders individually.

dmitrymyl commented on spoc/models/dataframe_models.py:
```diff
@@ -182,11 +275,43 @@ def validate_header(self, data_frame: DataFrame) -> None:
+        """Returns the position fields as a dictionary
+        of framgent index to the respective fields"""
+        return {
+            i: ["chrom", f"start_{i}"] for i in range(1, self._number_fragments + 1)
```
Okay, this works only for intrachromosomal pixels. If we want to have also interchromosomal ones, we've got to change smth. But for now it might work and we could schedule changes later.

Mittmich replied:
Good catch! Added the handling of interchromosomal contacts.

dmitrymyl commented on spoc/models/dataframe_models.py:
```diff
@@ -182,11 +275,43 @@ def validate_header(self, data_frame: DataFrame) -> None:
         """
+        self.validate_header(data_frame)
+        if isinstance(data_frame, duckdb.DuckDBPyRelation):
+            # duckdb does not support schema validation
```
Is it applicable for intermediate query results? If no, i.e. only works with parquet files, we could use `pyarrow.parquet` for [schema validation](https://stackoverflow.com/questions/76066470/how-to-show-user-schema-in-a-parquet-file-using-duckdb).

Duckdb precalculates schemas at each step anyway, so we might retrieve this info somehow.

Otherwise, we won't be able to validate schemas of intermediate results, and will have to check materialized dataframes at the end of queries.

Mittmich replied:
The shape of the schema is validated in `validate_header` for both  `duckdb.DuckDBPyRelation` and `pd.DataFrame`, the extended pandera validation, however is only done for `pd.DataFrmae`. That onle includes types in this case. I think for passing validating query steps, validation of headers should be enough.

dmitrymyl replied:
Okay, there might be some quirks with no intermediate checking of columns, but they would be rare.

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -0,0 +1,254 @@
+from spoc.io import DUCKDB_CONNECTION
+
+
+class GenomicData(Protocol):
```
Why aren't here any returns in methods? Maybe I don't understand smth, like if it's some typing object.

Mittmich replied:
Indeed, that is a protocol definition (https://peps.python.org/pep-0544/). It allows structural typing, whereby we don't need specific inheritance, but objects can implement a protocol and we can pass the protocol as a type for that method.

dmitrymyl replied:
Cool!

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -0,0 +1,254 @@
+        anchors (Optional[List[int]]): The list of anchor values (optional).
+    """
+
+    mode: str
```
No description of allowed modes here...

Mittmich replied:
Added description

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -0,0 +1,254 @@
+        """Returns the result as a dataframe object, either in memory or as a relation object"""
+        return self._data
+
+    def load_result(self) -> pd.DataFrame:
```
Oh, does this imply an implementation of two backends for the query engine or an application of the query engine to data preprocessing?

Mittmich replied:
Good, catch, this was actually a remonant form some early debugging. Removed the dask reference.

dmitrymyl commented on spoc/query_engine.py:
As far as I understand, @Mittmich provided stubs of query steps and implemented one of them, Snipper.

Mittmich replied:
Yes, exactly.

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -0,0 +1,254 @@
+        return self._schema
+
+
+class BasicQuery:
```
Looks sensible -- a generic code and everything depends on how individual steps are implemented.

Mittmich replied:
Indeed

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -0,0 +1,254 @@
+        """Return the schema of the underlying data"""
+
+
+class QueryStep(Protocol):
```
Shouldn't all query steps (like Snipper) inherit from this class?

dmitrymyl replied:
Or maybe it's some typing trick? :)

Mittmich replied:
Indeed, see other comment :)

Mittmich replied:
This is a very good explanation about nominal/structural typing: https://www.youtube.com/watch?v=xvb5hGLoK0A&t=1080s

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -0,0 +1,254 @@
+        """Validate the filter against the data schema"""
+        # check whether an anchor is specified that is not in the data
+        if self._anchor_mode.anchors is not None:
+            if not all(
```
Smth complicated... Are anchors numeric values?

Mittmich replied:
Yes, the refer to the positions as index.

dmitrymyl commented on spoc/io.py:
```diff
@@ -17,6 +18,9 @@
 from spoc.fragments import Fragments
 
+# Instantiate one duckdb connection to be used for all duckdb relations
+DUCKDB_CONNECTION = duckdb.connect(database=":memory:")
```
Might it cause any problems with memory consumption? I don't know how duckdb stores executed queries, so I fear it might clog and some garbage collection might be needed.

So, there are two options:
1. Singleton db connection.
    1. Pros: probably can utilize dataframes already loaded into duckdb for other queries.
    2. Cons: might clog memory with repetitive loading of dataframes and queries.
2. Instantiate db connection for each query separately.
    1. Pros: predictable behaviour regarding RAM (all data is loaded at the beginning of the query and cleaned upon the ending of the query and closing the connection).
    2. Cons: can't use potentially loaded dataframes before the query.

I am not sure that the architecture of our query engine allows to reuse dataframes previously uploaded to duckdb (to save time on repeated conversion from pandas to Arrow). I am also not sure how duckdb deals with memory and if it does GC. But I can expect some weird behaviour with memory, so I would suggest to instantiate db connection for each query separately.

dmitrymyl replied:
This solution would be problematic with lazy nature of our query engine (potentially).

Mittmich replied:
I don't think so to be honest. As everything is a connection in memore from a duckdb perspective, I think the best way is to let duckdb manage when to load what. The bulk of the operations is done anyway on  `duckdb.DuckDBPyRelation` objects, which are not loaded. Loading is only done when we call `df()`, which then is handed back as a pandas object and duckdb frees that memory again. Duckdb only recommends multiple connections, when threading/multiprocessing is done outside of duckdb.

dmitrymyl replied:
Okay, let's go with this

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -0,0 +1,254 @@
+        """
+        if isinstance(data, dd.DataFrame):
+            data = data.compute()
+        return duckdb.from_df(data, connection=DUCKDB_CONNECTION)
```
Duckdb might not like converting the same dataframe into Arrow multiple times and throw an error or make a duplicate of the DF in memory. Do we have data on how exactly it will work?

Mittmich replied:
Indeed, this method is only intended to be used with `pd.DataFrame` or `dd.DataFrame`. Fixed the type signature.

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -0,0 +1,254 @@
+            subset_positions = list(position_fields.values())
+        for fields in subset_positions:
+            # check two or three-way definition of fields
+            if len(fields) == 3:
```
I understand this case: it is when fragment coordinates in Pixels/Contacts have chrom, start, end. Then, genomic query is rather simple as described in `output_string` below. Right?

Mittmich replied:
Indeed

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -0,0 +1,254 @@
+                output_string = f"""(data.{chrom} = regions.chrom and
+                                     (data.{start} between regions.start and regions.end or 
+                                       data.{end} between regions.start and regions.end))"""
+            else:
```
This case is a bit trickier. Is it for binned data like Pixels, where we store only start coordinates so far? If so, then the `output_string` is incorrect, because `data.start` can be upstream of `regions`, but `data.start + bin_size` can be between `regions.start` and `regions.end`.

To mitigate that, in [my implementation of snipping](https://github.com/gerlichlab/spoc/blob/a17e2f1b10059935636bd16b198742f9d41f3642/spoc/snipping/snipping_strategies.py#L139) I extended `regions.start` by subtracting `bin_size` from it. That was done specifically for snipping triplets in a specific manner, so that solution worked there.

For spoc I would suggest to branch based on binning status of the dataset. If it is binned, then each fragment has only chrom and start, while end is constructed dynamically by `end AS start + {bin_size}`. And then we do `output_string` as
```{python}
output_string = f"""(data.{chrom} = regions.chrom and
                                     (data.{start} between regions.start and regions.end or 
                                       data.{end} between regions.start and regions.end))"""
```

If it is not binned, then each fragment has chrom, start, end. We don't need to modify anything and just go through to `output_string`.

IMO it is better to make a virtual end column for binned dataset rather than modify regions, since we provide the common interface of multi-way contacts for the query engine this way.

dmitrymyl replied:
Or maybe I am wrong and snipping of bins is defined exactly like this...

dmitrymyl replied:
I see, you changed that in `duckdb-transform-aggregation`, nice!

Mittmich replied:
Indeed, fixed it there :)

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -0,0 +1,254 @@
+        df = genomic_df.set_alias("data").join(
+            regions.set_alias("regions"), self._contstruct_filter(position_fields)
+        )
+        return QueryResult(df, self._get_transformed_schema(input_schema))
```
Transformed schema is needed only for subsequent queries in spoc and doesn't do anything with duckdb, right?

Mittmich replied:
Exactly, it just propagates the planned changes.

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -0,0 +1,254 @@
+    def __repr__(self) -> str:
+        return f"Snipper(anchor_mode={self._anchor_mode})"
+
+    def __call__(self, genomic_data: GenomicData) -> GenomicData:
```
I like this method, it's expressive.

Mittmich replied:
Thanks :)

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -0,0 +1,254 @@
+        return self.__repr__()
+
+
+class Snipper:
```
So, this is `SELECT` in my original nomenclature, right?

Mittmich replied:
Exactly

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -0,0 +1,254 @@
+        # instantiate query result
+        query_result = QueryResult(input_data.data, input_data.get_schema())
+        # run query
+        for step in self._query_plan:
```
I like this loop

Mittmich replied:
Thanks!

dmitrymyl commented:
@Mittmich I think I checked all the files except for tests and would be happy to discuss your code and my comments :) I also checked out the `duckdb-transform-aggregation` branch and noticed that you fixed some of my comments there. Maybe it will make sense to open a PR for merging it into this branch first and then merge the result into main? I will add then comments first to `duckdb-transform-aggregation` and then we could get a view of the more complete query engine.

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -0,0 +1,254 @@
+
+
+# TODO: think about allowing anchor composition
+class Anchor(BaseModel):
```
Anchor specifies which fragments to select in Contacts/Pixels, right? To me it sounds not idiomatic enough. I think that smth idiomatic would be `Fragments` or `FragmentSelector`, but that is too long to type.

Mittmich replied:
Yeah, I agree, anchor is somewhat abstract. Maybe let's unify when we implement the MultiAnchor we talked about?

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -0,0 +1,254 @@
+            # check two or three-way definition of fields
+            if len(fields) == 3:
+                chrom, start, end = fields
+                output_string = f"""(data.{chrom} = regions.chrom and
```
This filter doesn't cover the case of nested genomic intervals, i.e. when a bin contains a region. However, it covers the case when a region contains a bin.

Mittmich replied:
No, it does not. It was somewhat by design to keep the interface simple. Do you expect a lot of use cases when regions are smaller than bins? Keep in mind that the regions here are intended to act like the windows that we defined for the snipping in cooltools to provide a view onto the genomes.

dmitrymyl replied:
I can imagine smth like this, e.g. find all multiway contacts that have one fragment overlapping a specific promoter (or a set of promoters, or CTCF sites). For our cases it might be not a frequent use case, but for general audience -- yeah. So I suggest we keep it stringent. Anyway, it's just adding another subclause `or regions.start between data.start and data.end`, shouldn't be too complicated.

I am curious, though, how does duckdb evaluate OR clauses? In python, if A from (A or B) is true, B is not evaluated. If duckdb does the same, that would be cool. So if we expect many overlaps, than most of the times in this case duckdb will evaluate only first two subclauses, so we shouldn't drop in performance much. But it would be interesting to test (monkey-patching?).

dmitrymyl replied:
Cool, so now we have only this comment left (and maybe arrows in the UML diagram, but that's not very important).

Mittmich replied:
Alright! I added the additional condition :) I am almost certain that duckdb does lazy evaluation of OR clauses, so I don't think there is a performance penalty!

dmitrymyl replied:
Yay, thanks!

Mittmich commented:
Hey @dmitrymyl,
thanks for the thorough review! Regarding your points:

# Naming suggestions

> 1. query = BasicQuery([...]); result = query.query(dataset). query.query sounds redundant. Maybe we should just implement __call__ to make it query(dataset)? Although, this stuff is a bit advanced python feature, less experienced users might not be familiar with it and might find it confusing.

I totally agree with this! This also ties into the naming of `BasicQuery` since it's a very fundamental object. I would be in favor of naming it `Query` and having a `build` method that is equivalent to the current `query` method?

> 2 .result.compute() sounds better and in line with dask than result.load_result().

I agree with this and will rename.

> 3. Snipper, Aggregation, Transformation — imo verbs would be shorter to type and they are also in line with SQL syntax (thus helping with adaption to our query engine). So I suggest to rename all QueryStep into verbs like Snip, Aggregate,

Sounds good! I already started to implement some of this in the `duckdb-transform-aggregation` branch, so this fits well.

> 4. Snipping naming

Yes, I also think that snipping is jargon and we should find something else. I don't like `Select` that much as - as you say -  has another meaning in SQL. I like `overlap` as a name. I would then also merge the single and multiregion capabilities into one class and make the behavior dependent on the passed input.

 > 5. In the docs, Filter along the rows — I have always found this terminology confusing in pandas. Does it subset rows or subset columns? I suspect the former. It could also be rewritten from the perspective of contacts and individual fragments, but this is not crucial ATM.

Yeah, this is meant to subset rows. We could also capture this by implementing a `RegionOffsetOverlap` to allow overlapping with relative regions that have been added already. Then we wouldn't need to talk about rows or columns, but concepts.

# Questions


> What are get_position_coordinates() of gDataSchema? It was not described in the UML diagram (maybe it will be clear after I check the code).

This is a method that returns the position coordinates from a GenomicDataSchema. This allows to validate schemas during construction of queries.

> Mermaid diagram: should all arrows be in reverse? It's the case with [UML diagrams](https://www.visual-paradigm.com/guide/uml-unified-modeling-language/uml-class-diagram-tutorial/), but maybe syntax is different here. Anyway, it doesn't matter for the project per se.

Hmm, at least for the realization relationships, the arrow points towards the class that realizes a protocol (https://www.ibm.com/docs/en/dma?topic=diagrams-realization-relationships), so I think that should be fine?

> It was a bit confusing that QueryResult is an output of the BasicQuery and implements the gDataProtocol at the same time. Initially, I was thinking that the QueryResult is an actual materialization, like a 2D pileup matrix. But it is actually a lazy evaluated result, not the data. Then it makes sense to say it implements the gDataProtocol.

Ah yes, maybe we can also rename the `QueryResult` to something like `QueryPlan`? Then we can rename the argument `query_plan` of `BasicQuery` (which we will rename as `Query`) to something like `query_steps`. Then the `Query` object will get a list of query steps and it's `build` method generates a `QueryPlan`, which has a `compute` method. What do you think?

> However, I am still not sure that the QueryResult should implement the gDataProtocol. It makes sense since the BasicQuery can accept the QueryResult. But what if the QueryResult actually relates to smth not being Pixels or Contacts, like 2D pileup matrix? What sense is there for it to follow the gDataProtocol? As an alternative, QueryResult might or might not implement the gDataProtocol to accomodate for this ambiguity.

The `GenomicDataProtocol` is quite flexible as it only contains methods to get aspects of the schema of the data. Whether that schema is compatible with other query steps depends on the specific query step and their validation method. So in theory, we can have 2d pixels, but basically, no query steps could accept it as input, which is fine since it would be more of a result.

> BasicQuery.query() builds the query plan for postponed execution. The query plan is specified in the class instance and is applied to a dataset. This implies that query steps are independent of the dataset. Can it be otherwise: that features of the dataset (shape, contact orders etc) dynamically dictate the specific choice of procedures (in a way that user doesn't know that beforehand and cannot write another query plan)?

Hmm, that is a good question! I am not sure whether this is the case... And the steps do have access to the schema that all the query steps produce as building the query plan involves pre-building the schema.

# Merging of branches

I think it's great that you already looked at the code in the `duckdb-transform-aggregation` branch! What I would propose is that I merge that branch here and we discuss all your comments together. Otherwise we will have many discussions that need to refer to the other branch.

Regarding your comments: I can't really see them, unfortunately, did you add them here?

dmitrymyl reviewed (changes requested):
Okay, with this comment I try to make my review comments visible...

dmitrymyl commented on spoc/io.py:
```diff
@@ -17,16 +20,21 @@
+            self._parquet_reader_func = partial(duckdb.read_parquet, connection=DUCKDB_CONNECTION)
+        elif data_mode == DataMode.DASK:
             self._parquet_reader_func = dd.read_parquet
         else:
```
Could it be that a user would submit an incorrect data mode? This clause won't catch that and reset to the default pandas backend.

Alternative would be `elif data_mode == DataMode.PANDAS: ... else: raise ValueError`.

dmitrymyl replied:
It might sound too pedantic, but it would be expressive to implement such structure of branching when evaluating the data mode everywhere... Although, for us it shouldn't matter much since we know that we have only 3 data modes, but for other readers of the code or us in the future it would be a nice explicit explanation.

Btw, python 3.10 supports `switch ... case`, but that would require upping the python version.

Mittmich replied:
Fair, I added the check for supported datamodes!

dmitrymyl reviewed (changes requested):
Cool! I added a couple of comments here as well.

dmitrymyl commented:
Hey @Mittmich, thanks for the replies and quick fixes!

# Naming suggestions
> > 1. query = BasicQuery([...]); result = query.query(dataset). query.query sounds redundant. Maybe we should just implement **call** to make it query(dataset)? Although, this stuff is a bit advanced python feature, less experienced users might not be familiar with it and might find it confusing.
> 
> I totally agree with this! This also ties into the naming of `BasicQuery` since it's a very fundamental object. I would be in favor of naming it `Query` and having a `build` method that is equivalent to the current `query` method?

Sounds very good!

> 
> > 2 .result.compute() sounds better and in line with dask than result.load_result().
> 
> I agree with this and will rename.

Great

> 
> > 3. Snipper, Aggregation, Transformation — imo verbs would be shorter to type and they are also in line with SQL syntax (thus helping with adaption to our query engine). So I suggest to rename all QueryStep into verbs like Snip, Aggregate,
> 
> Sounds good! I already started to implement some of this in the `duckdb-transform-aggregation` branch, so this fits well.
> 

Perfect

> > 4. Snipping naming
> 
> Yes, I also think that snipping is jargon and we should find something else. I don't like `Select` that much as - as you say - has another meaning in SQL. I like `overlap` as a name. I would then also merge the single and multiregion capabilities into one class and make the behavior dependent on the passed input.

`Overlap` and `MultiOverlap` sound good, I also like them.

> 
> > 5. In the docs, Filter along the rows — I have always found this terminology confusing in pandas. Does it subset rows or subset columns? I suspect the former. It could also be rewritten from the perspective of contacts and individual fragments, but this is not crucial ATM.
> 
> Yeah, this is meant to subset rows. We could also capture this by implementing a `RegionOffsetOverlap` to allow overlapping with relative regions that have been added already. Then we wouldn't need to talk about rows or columns, but concepts.
> 

`RegionOffsetOverlap` sounds interesting, this might be a nice feature, but let's think about it later :)

# Questions
> > What are get_position_coordinates() of gDataSchema? It was not described in the UML diagram (maybe it will be clear after I check the code).
> 
> This is a method that returns the position coordinates from a GenomicDataSchema. This allows to validate schemas during construction of queries.

Got it

> 
> > Mermaid diagram: should all arrows be in reverse? It's the case with [UML diagrams](https://www.visual-paradigm.com/guide/uml-unified-modeling-language/uml-class-diagram-tutorial/), but maybe syntax is different here. Anyway, it doesn't matter for the project per se.
> 
> Hmm, at least for the realization relationships, the arrow points towards the class that realizes a protocol (https://www.ibm.com/docs/en/dma?topic=diagrams-realization-relationships), so I think that should be fine?

I think that in the link you sent the arrow points to the class that specifies the protocol...

> 
> > It was a bit confusing that QueryResult is an output of the BasicQuery and implements the gDataProtocol at the same time. Initially, I was thinking that the QueryResult is an actual materialization, like a 2D pileup matrix. But it is actually a lazy evaluated result, not the data. Then it makes sense to say it implements the gDataProtocol.
> 
> Ah yes, maybe we can also rename the `QueryResult` to something like `QueryPlan`? Then we can rename the argument `query_plan` of `BasicQuery` (which we will rename as `Query`) to something like `query_steps`. Then the `Query` object will get a list of query steps and it's `build` method generates a `QueryPlan`, which has a `compute` method. What do you think?

That actually sounds cool!

> 
> > However, I am still not sure that the QueryResult should implement the gDataProtocol. It makes sense since the BasicQuery can accept the QueryResult. But what if the QueryResult actually relates to smth not being Pixels or Contacts, like 2D pileup matrix? What sense is there for it to follow the gDataProtocol? As an alternative, QueryResult might or might not implement the gDataProtocol to accomodate for this ambiguity.
> 
> The `GenomicDataProtocol` is quite flexible as it only contains methods to get aspects of the schema of the data. Whether that schema is compatible with other query steps depends on the specific query step and their validation method. So in theory, we can have 2d pixels, but basically, no query steps could accept it as input, which is fine since it would be more of a result.

Okay, I got it.

> 
> > BasicQuery.query() builds the query plan for postponed execution. The query plan is specified in the class instance and is applied to a dataset. This implies that query steps are independent of the dataset. Can it be otherwise: that features of the dataset (shape, contact orders etc) dynamically dictate the specific choice of procedures (in a way that user doesn't know that beforehand and cannot write another query plan)?
> 
> Hmm, that is a good question! I am not sure whether this is the case... And the steps do have access to the schema that all the query steps produce as building the query plan involves pre-building the schema.

Okay, let's think that this is not the case. But maybe we should write that down somewhere to not forget this possibility.


# Merging of branches
> I think it's great that you already looked at the code in the `duckdb-transform-aggregation` branch! What I would propose is that I merge that branch here and we discuss all your comments together. Otherwise we will have many discussions that need to refer to the other branch.
> 

Yeah, that'll be convenient

dmitrymyl reviewed (changes requested):
Nice!

Mittmich commented:
Alright, I merged the other branch and renamed the classes as discussed :) Would be great if you could also have a look at the other functionality (Described in the query_engine_usage.ipynb notebook)! Thanks for the help!

dmitrymyl commented on spoc/contacts.py:
```diff
@@ -1,15 +1,22 @@
-from typing import List, Optional, Dict
-import pandas as pd
+
+from itertools import permutations
```
What's the reason for splitting imports? :)

Mittmich replied:
It's done by reorder-python-imports in the pre-commit hooks. It separates builtin from third-party from imports in this repo. It's very handy because it reorders imports automatically :)

dmitrymyl replied:
Ah, I meant that before there were multiple objects imported per line (List, Optional, Dict), and now they are imported in individual lines &mdash; but if that's an automatic behaviour, then it doesn't matter :)

dmitrymyl commented on spoc/contacts.py:
```diff
@@ -387,7 +394,7 @@ def _sort_chromosomes(self, df: DataFrame, number_fragments: int) -> DataFrame:
     def _generate_binary_label_mapping(
         self, label_values: List[str], number_fragments: int
-    ) -> Dict[str, str]:
+    ) -> Dict[Tuple[str, ...], Tuple[str, ...]]:
```
Am I correct that the behaviour of the method didn't change, but only the type of the output got clarified?

Mittmich replied:
Yes, indeed. I added mypy checks to the pre-commit hooks and it found this mistake :)

dmitrymyl commented on spoc/models/dataframe_models.py:
```diff
@@ -59,20 +65,33 @@ def get_contact_order(self) -> int:
+    def get_binsize(self) -> Optional[int]:
+        """Returns the binsize of the genomic data"""
+
+    def get_region_number(self) -> Optional[int]:
```
Is it a number of unique regions in Contacts+Regions?

Mittmich replied:
Indeed, the number of unique regions if present, otherwise none

dmitrymyl commented on spoc/pixels.py:
```diff
@@ -198,43 +201,45 @@ class GenomicBinner:
-    def _get_assigned_bin_output_structure(self):
-        columns = [f"chrom_{index}" for index in range(1, self._contact_order + 1)] + [
-            f"start_{index}" for index in range(1, self._contact_order + 1)
+    def _get_assigned_bin_output_structure(self, contact_order: int):
```
Was this change done because there was no proper assignment of `self._contact_order`?

Mittmich replied:
yes, this was to solve a mypy problem that `self._contact_order` could be `None` and that if we pass it to this method the type checker fails, because it's not smart enough to realize that everywhere we use this method, it will be not None. In any case, i think it's cleaner to explicitely pass the contact_order

dmitrymyl commented on .pre-commit-config.yaml:
Does it work on devs side (e.g., in vscode) or on github side for PRs?

Mittmich replied:
It works on dev side: you can add this to git as pre-commit hook and git will run it for you for each commit (and block commits if one of the checks fails), which is nice to build these tools into your workflow. You can also run it by hand using the `pre-commit` cli.

dmitrymyl commented on docs/query_engine_interface.md:
```diff
@@ -69,131 +73,117 @@ classDiagram
 
-query2 = BasicQuery(query_plan=query_plan_2)
+result = Query(query_plan_aggregate)\
+                  .query(cis_filtered)\
```
`query_plan_aggregate` contains `trans_filtered`, but here we query `cis_filtered`. Is it a typo or a logic that I missed?

Mittmich replied:
Indeed, the query plan conatins a step that concatenates the passed genomic data with the `trans_filtered` genomic data. So we build the queyr plan for `cis_filtered`, which includes concatenating it with `trans_filtered`.

dmitrymyl replied:
I see, wow. Not straightforward, but okay :) What if we would need to concatenate more than 2 gData? I would expect then `Query.query` to accept a collection of gData, so the type of the input arg would be `Union[gDataProtocol, List[gDataProtocol]]`.

dmitrymyl replied:
Alternatively, it can be consequent `Concatenate` for each dataset.

dmitrymyl commented on docs/query_engine_interface.md:
```diff
@@ -69,131 +73,117 @@ classDiagram
-query2 = BasicQuery(query_plan=query_plan_2)
+result = Query(query_plan_aggregate)\
+                  .query(cis_filtered)\
+                  .load_result()
```
Should it be `.compute()` now?

Mittmich replied:
Indeed, fixed it.

dmitrymyl commented on docs/query_engine_usage.md:
```diff
@@ -18,9 +18,9 @@ The most important ingredient in this query language is a class that implements
 
 This way, query steps can be combined into a query plan that specifies the analysis to be executed. Specific examples of query steps are:
```
Maybe we could describe general types of steps (Overlap, Transform, Aggregate) to show our framework and then give examples of each step?

dmitrymyl commented on docs/query_engine_usage.md:
````diff
@@ -63,42 +63,42 @@ target_region = pd.DataFrame({
+query = Query(query_steps=query_steps)
 ```
 
 The `.query` method executes the query plan and retuns a `QueryResult` object
````
Hasn't been edited: `.build` and `QueryPlan`

dmitrymyl commented on docs/query_engine_usage.md:
````diff
@@ -349,3 +349,386 @@ BasicQuery(query_plan=query_plan)\
+```python
+query_steps = [
+    Overlap(target_regions, anchor_mode=Anchor(mode="ANY")),
+    RegionOffsetTransformation(),
````
`OffsetMode` mentioned in the paragraph above but no example here.

Mittmich replied:
Made default explicit

dmitrymyl commented on docs/query_engine_usage.md:
```diff
@@ -349,3 +349,386 @@ BasicQuery(query_plan=query_plan)\
+query_plan = [
+    Overlap(target_regions, anchor_mode=Anchor(mode="ANY")),
+    RegionOffsetTransformation(),
+    OffsetAggregation('count'),
```
`'count'` is the column which we aggregate, right? Then here is no example of `AggregationFunction`.

Mittmich replied:
Indeed. I added the parameters to make it clearer.

dmitrymyl commented on docs/query_engine_usage.md:
```diff
@@ -349,3 +349,386 @@ BasicQuery(query_plan=query_plan)\
 In this example, the contact overlapping both regions is duplicated.
```
This is due to region IDs, right? I am curious, if we would need a type of output when there is no regions and we need only unique contacts. This can easily be done with pandas (drop columns, drop duplicates), so I don't think we have to implement that in the query engine for now.

Mittmich replied:
Indeed, it is because the there are regions that are overlapping and we therefore report contacts that fall within both regions twice. I think it is a good idea to determine whether we want that behavior. One could imagine to adding it as a parameter to the `Overlap` class, perhaps as we described in the concept for multioverlap:
```
    MultiOverlap(
        regions=[loop_bases_left, loop_bases_right], # Regions to overlap
        add_overlap_columns=False, # Whether to add the regions as additional columns
        anchor=MultiAnchor(
            fragment_mode="ANY",
            region_mode="ANY", positions=[0,1]
        ), # At least one fragment needs to overlap at least one region
    ),
```
This way we could incorporate it in query plans and wouldn't need to resort to pandas to get rid of duplicates.

dmitrymyl replied:
Yeah, it might work this way :)

dmitrymyl commented on docs/query_engine_usage.md:
````diff
@@ -349,3 +349,386 @@ BasicQuery(query_plan=query_plan)\
+
+```python
+query_plan = [
+    Overlap(target_regions, anchor_mode=Anchor(mode="ANY")),
````
I would use `mode="ALL"` for pileups. Besides, it would be a nice opportunity to show case this mode.

dmitrymyl commented on docs/query_engine_usage.md:
```diff
@@ -349,3 +349,386 @@ BasicQuery(query_plan=query_plan)\
+query_plan = [
+    Overlap(target_regions, anchor_mode=Anchor(mode="ANY")),
+    RegionOffsetTransformation(),
+    OffsetAggregation('count'),
```
Also, it is unclear, over which columns we aggregate. Well, based on the class name, I can deduce that we aggregate over `["offset_1", "offset_2", "offset_3"]` for triplets, but this should maybe be clarified somewhere. Also, is it possible to aggregate over a subset of offset columns?

Mittmich replied:
Good point! My reasoning was that `OffsetAggregation` would always aggregate over all offset columns and we would have a separate aggregation for reducing along an offset dimension. But indeed, there is actually no difference other than the function we would use. I think also densify output would work for that task. I would then implement subsets!

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -172,83 +199,367 @@ def __call__(self, genomic_data: GenomicData) -> GenomicData:
+        )
+
 
+class AggregationFunction(Enum):
```
Wouldn't it be too wordy to specify a non-default agg function in `OffsetAggregtaion`?

Mittmich replied:
I would implement that as a separate type that can be passed as a function, for example, a basic implementation would be that we allow `Union[AggregationFunction|str]` for the function argument and if it's a string, we call `.format` with the value column. So if you want something like `MIN(column)`, you would pass `MIN({})`.

dmitrymyl replied:
Okay, that sounds reasonable for custom functions. I was a bit confused that if I want to specify a non-default function (like SUM), I will have to type `AggregationFunction.SUM`, which is too long. But then we can come up with an alias, like `import torch.nn.functional as F`.

For the case that you described, passing smth like `MIN({})` seems to be a bit redundant -- we can attach `({})` under the hood.

Speaking of custom functions, we have two sorts of them: defined in duckdb and UDFs. Former should be simple to use, while I am curious how UDFs are defined and used in duckdb (since one can write a duckdb UDF in python, this seems to be a nice feature for the future).

Mittmich replied:
Inded, UDFs would be nice in the future, but would leave that for now

dmitrymyl reviewed (changes requested):
@Mittmich I checked new code except for tests and implementations of RegionOffsetTransformation and OffsetAggregation (will do later). I added my comments to code and some general notes below:

Now the query engine emerges as a powerful tool! I think now we have a good example of how we can use it (with pileups). I am curious how easy it will be to add new steps to it -- maybe some contribution docs in the (not so near) future?

1. I rarely see enums in public API of scientific packages, so they feel inconvenient to me. I would design their functionality as str kwarg with nested ifs, but then there is a problem of repetitive code and unclear specifications.
2. There are no performance queries like setting numbers of cores and memory limit. These params could be kwargs for the Query class.
3. Offset coordinates are added separately, interesting implementation, but makes sense for downstream steps.
4. Do we have support for the normalized data? Like `"normalized_count"` or smth.
5. RegionOffsetTransformation and OffsetAggregation are nouns, should we switch to verbs? But it will be lengthy anyway...
6. Should we maybe make "stub" classes for Transform and Aggregate and inherit from them to establish the nomenclature of steps in the codebase?

I feel like I should update the nomenclature of query steps in my doc in Teams so that we could check which steps we have already implemented and which are left.

Otherwise looks great, thanks as usual!

Mittmich commented:
Hi @dmitrymyl , thanks for the review! I addressed your comments :)

Regarding your points:

> 1. I rarely see enums in public API of scientific packages, so they feel inconvenient to me. I would design their functionality as str kwarg with nested ifs, but then there is a problem of repetitive code and unclear specifications.

I agree! I added the possibility to instantiate the classes with string arguments :) 

> 2. There are no performance queries like setting numbers of cores and memory limit. These params could be kwargs for the Query class.

Good point, added a trello card: https://trello.com/c/gPAAy9Um/95-add-performance-setting-queries-to-query-class

> 3. Do we have support for the normalized data? Like "normalized_count" or smth.

Yes,  one just needs to pass it as `value_column` for the `OffsetAggregations`.

> 4. RegionOffsetTransformation and OffsetAggregation are nouns, should we switch to verbs? But it will be lengthy anyway...

Hmm, indeed, but I would wait with renaming until we have a few example, because we may need to group things together again anyway.

> 6. Should we maybe make "stub" classes for Transform and Aggregate and inherit from them to establish the nomenclature of steps in the codebase?

I thought about it, but I wouldn't make stub classes until we have specific separation of behavior as at the moment they all have the same interface, namely the `QueryStepProtocol`. I actually think it's good to have more conceptual separation, but similar behavior in terms of code.

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -172,83 +218,375 @@ def __call__(self, genomic_data: GenomicData) -> GenomicData:
+        offset_combinations = pd.DataFrame(
+            product(
+                np.arange(
+                    -(np.floor(windowsize / int_binsize) * int_binsize),
```
This line seems like a bit magic to me... Should we even do it if all gData has start and end coordinates?

dmitrymyl replied:
Okay, this is empty dataset with binned offsets which will be used later to join with [data x regions + offsets].

Mittmich replied:
Indeed :)

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -172,83 +218,375 @@ def __call__(self, genomic_data: GenomicData) -> GenomicData:
+            raise ValueError("No binsize specified in data schema.")
+        int_binsize: int = binsize
+        # create combinations of offsets
+        offset_combinations = pd.DataFrame(
```
`offset_combinations` shouldn't be a heavy object, right? It seems like the amount of rows would be `(2 * (window_size // bin_size)) ** order` &mdash; feels heavy for contacts of very high orders. But I don't think we will use offsets for orders of more than 4, so no need to optimize for now.

Alternatively, this can be constructed in duckdb...

Mittmich replied:
No, I don't think so either. If it becomes a problem, we can construct it in duckdb :)

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -172,83 +218,375 @@ def __call__(self, genomic_data: GenomicData) -> GenomicData:
+            binsize=input_schema.get_binsize(),
+        )
+
+    def _aggregate_offsets(
```
Shouldn't be somewhere here a groupby statement?

Mittmich replied:
It's in the `.aggregate` method of duckdb

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -172,83 +218,375 @@ def __call__(self, genomic_data: GenomicData) -> GenomicData:
+            .aggregate(
+                ",".join(offset_columns + [aggregation_string]),
+            )
+            .order(",".join(offset_columns))
```
This can be costly in case there are too many offset coordinates, but it shouldn't happen in our use cases.

Mittmich replied:
Indeed, but i think we might not have the situation where hwe have a large number of offsetcoordinates

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -172,83 +218,375 @@ def __call__(self, genomic_data: GenomicData) -> GenomicData:
-    """Filter for filtering mapped regions
-    of contacts or pixels."""
+    SUM: str = "SUM"
+    AVG_WITH_EMPTY: str = "AVG_WITH_EMPTY"
```
I was thinking about following the numpy naming scheme with `mean` and `nanmean`, but then there is no `MEAN` alias for `AVG` in duckdb...

Mittmich replied:
Hmm, I wouldn't do that since `AVG_WITH_EMPTU` behaves different from `nanmean` the former fills in the empty values, whereas the latter ignores them.

dmitrymyl replied:
True!

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -172,83 +218,375 @@ def __call__(self, genomic_data: GenomicData) -> GenomicData:
+        # get offset columns
+        offset_columns = [f"offset_{i}" for i in position_fields.keys()]
+        # construct join and coalesce output
+        data_frame = (
```
That's a bit confusing query (because I am not familiar with `project` and `COALESCE`) &mdash; I will parse it later.

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -172,83 +218,375 @@ def __call__(self, genomic_data: GenomicData) -> GenomicData:
+        )
+        if self._densify_output:
+            # Take maximum windowsize of regions
+            windowsize = (
```
I was surprised by the fact that `windowsize` is estimated here instead of being an argument. But then I was expecting it to be an argument in `RegionOffsetTransformation`, and it's not there as well...

dmitrymyl replied:
Okay, now terminology is a bit confusing, since it's changed compared with our previous implementation of triplet pileups.

Here, offsets &mdash; are distances from each genomic region to each fragment, right?

Then window size is a confusing term here. How genomic regions look like? Should I already extend them +- window size before submitting to the query engine?

Mittmich replied:
Windowsize is now an explicit argument to `Overlap` :)

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -172,83 +218,375 @@ def __call__(self, genomic_data: GenomicData) -> GenomicData:
+        )
+
+
+class OffsetMode(Enum):
```
If regions submitted to RegionOffsetTransformation already have been extended +- window size, why would a user need to calculate different offsets except for `MIDPOINT`?

dmitrymyl replied:
Are there other potential use cases?

dmitrymyl replied:
Aaah, the OffsetMode is for fragments, not for regions! Okay, this works.

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -172,83 +218,375 @@ def __call__(self, genomic_data: GenomicData) -> GenomicData:
+            data_frame.set_alias("data")
+            .join(
+                empty_dense_output.set_alias("empty_dense_output"),
+                ",".join(offset_columns),
```
If distances (offsets) from regions to fragments are in nts (not binned), how the join with binned empty offset df works?

Mittmich replied:
See answer to comment, but in short, no onctacts allowed for no

dmitrymyl reviewed (changes requested):
Hi @Mittmich, thanks for the new commits!

It looks good and well designed, but I got lost in namings and implementations of RegionOffsetTransformation and OffsetAggregation. They represent necessary blocks for the pileup procedure, so I tried to understand how the procedure should go.

1. ROIs are transformed outside of the query engine:
    1. midpoint = (start + end) // 2
    2. start = midpoint - windowsize
    3. end = midpoint + windowsize
2. These transformed ROIs are submitted to `Overlap` and they are overlapped with `Pixels`.
3. Offsets -- are distances from midpoints of ROIs to the specified coordinate of each pixel bin (start, end, midpoint, both -- defined in Offset enum), by default it's start.
4. `RegionOffsetTransformation` calculates those offsets.
5. `OffsetAggregation` binnifies offsets (implicitly), selects the window size (relying on the fact that the ROIs are windowed midpoints) and aggregates counts of pixels by binned offsets if they fall within the window.

Does this flow reflect the one implemented?

I have some questions and suggestions:
1. I got lost with the "Offset" term (thinking past to our previous implementation of triplet pileups). Maybe we could rename it to "Distance"?
2. I couldn't pinpoint the binning of offsets &mdash; is it not present in the code and done implicitly relying on the fact that pixels are binned already?
3. Maybe we could instead introduce an operation `OffsetBinningTransformation`, that converts calculated offsets in nts into bins relative to midpoints of ROIs? This operation would be placed between `RegionOffsetTransformation` and `OffsetAggregation` thus removing implicit casting of real offsets into binned offsets (which relies on binning of pixels for now). Then `RegionOffsetTransformation` just calculates distances from fragments to midpoints of ROIs (obtained after Overlap with windowed ROIs) and could be applied to both pixels and contacts, the logic of casting real offsets into binned offsets is implemented in `OffsetBinningTransformation`, and aggregation in `OffsetAggregation`.
4. I haven't parsed how exactly `RegionOffsetTransformation` calculates offsets (=distances); but can it be implemented in `Overlap` as well similar to bioframe and bedtools?
5. I also think that it might be better to pass the window size explicitly somewhere. Like, for example, it can be `OffsetWindowFilter` or just `OffsetFilter` right after the `OffsetAggregation` to filter out rows where offsets are outside of (-windowsize, +windowsize).


I am curious what you think about it 🙂

dmitrymyl commented:
Regarding your replies to my replies for the first portion of the review, @Mittmich, everything sounds great, I agree with your points!

I have some other feature requests:
1. Can we implement a `.to_sql` method in `QueryPlan`? This would be very useful for development and debugging later.
2. If we pursue implementing pileups, it would be very nice to have an option of strand-specific pileups, i.e. before aggregation flip the sign of offset coordinates for those pixels that have a region with "-" strand. This can be implemented as a separate basic block, but then RegionSchema should be able to take "strand" (or another user-defined field) as an optional column. We can implement it later anyway, not in this PR, since we have enough stuff to deal with for now.

Mittmich commented:
Hey @dmitrymyl ,
Thanks for the thorough review! Regarding the more general points:

> Does this flow reflect the one implemented?

Indeed, that flow is the one implemented, with one exception: `OffsetAggregation` does not perform binning unless this is done with the offset columns. I agree however that this is confusing behavior and would propose the following:

- `OffsetAggregation` only operates on pixels and explicitly checks whether the passed data are pixels (This is done implicitly at the moment as contacts don't have a value column). 
- This would mean that to aggregate contacts by offset, we would first need to make pixels out of them, which to me seems fair enough. 

I think this will also answer a lot of the code comments that you made.

> 1. I got lost with the "Offset" term (thinking past to our previous implementation of triplet pileups). Maybe we could rename it to "Distance"?

Distance sounds good, will rename it.

> 2. I couldn't pinpoint the binning of offsets — is it not present in the code and done implicitly relying on the fact that pixels are binned already?

Indeed, see the comment above.


> 3. Maybe we could instead introduce an operation OffsetBinningTransformation, that converts calculated offsets in nts into bins relative to midpoints of ROIs? This operation would be placed between RegionOffsetTransformation and OffsetAggregation thus removing implicit casting of real offsets into binned offsets (which relies on binning of pixels for now). Then RegionOffsetTransformation just calculates distances from fragments to midpoints of ROIs (obtained after Overlap with windowed ROIs) and could be applied to both pixels and contacts, the logic of casting real offsets into binned offsets is implemented in OffsetBinningTransformation, and aggregation in OffsetAggregation.

Eventually, that would be great, but I think currently, the flow would be to have clients of the code go through creating pixels.

>  haven't parsed how exactly RegionOffsetTransformation calculates offsets (=distances); but can it be implemented in Overlap as well similar to bioframe and bedtools?

It could be, but I like the separation and it's a nice example of a transformation, I think.

> I also think that it might be better to pass the window size explicitly somewhere. Like, for example, it can be OffsetWindowFilter or just OffsetFilter right after the OffsetAggregation to filter out rows where offsets are outside of (-windowsize, +windowsize).

Yes, implicit windowsize handling is a problem. How about one can optionally pass window sizes and then regions are expanded to that size? And then it can be stored in the schema so we don't really on implicit calculations.

Mittmich commented:
Hey! I implemented all the suggestions above, except for the rename to Distance, I thought about it some more and I think offset is more intuitive. We could name the thing we historically called offset also `Anchor`? But I don't feel super strongly about it, happy to be convinced!

dmitrymyl commented on spoc/models/dataframe_models.py:
```diff
@@ -86,12 +90,14 @@ def __init__(
         contact_order: int,
         binsize: Optional[int] = None,
         region_number: Optional[int] = None,
+        window_size: Optional[int] = None,
```
The window size is specific to only 1 or 2 workflows for now. Should it be defined in `__init__`, even as Optional?

But I can't imagine a better solution for now. Put it into `**kwargs` and then add all of them to a `self.optional: Dict = kwargs` or to `self.__dict__.update(kwargs)`? This would make stuff implicit, which we want to avoid.

Mittmich replied:
Yeah, I think we could group the parameters into different modules, but tbh, this feels like overengineering

dmitrymyl commented on spoc/models/dataframe_models.py:
```diff
@@ -72,6 +72,10 @@ def get_region_number(self) -> Optional[int]:
         """Returns the number of regions in the genomic data
         if present."""
 
+    def get_window_size(self) -> Optional[int]:
```
The window size is a property of regions, not gData, right? I understand that we won't be able to have valid schemas for query steps otherwise, but this feels hardcoded for 1-2 workflows.

Which might be okay for now, we could solve this problem in another PR later.

Mittmich replied:
Yeah, it's true. At the moment `GenomicDataSchema` carries all information about the accumulating data. In the end, we need such an object to perform validation. I don't really have an idea about how to do it otherwise. I think it's fair in principle since in a sense the genomicdata that we accumulate are tied to a specific windowsize. It has been imparted by overlapping the region, but the result is linked to it.

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -84,22 +84,46 @@ class Overlap:
+                expanded_regions["start"] + expanded_regions["end"]
+            ) // 2
+            # expand regions
+            expanded_regions["start"] = expanded_regions["midpoint"] - window_size // 2
```
Wouldn't it be easier to have midpoint ± half_window_size? I think that people usually talk like that.

Mittmich replied:
Indeed, renamed!

dmitrymyl commented on spoc/query_engine.py:
```diff
@@ -293,10 +318,12 @@ def validate(self, data_schema: GenomicDataSchema) -> None:
-                "No binsize specified in data schema. This is required for densifying the output."
-            )
+        # check for binsize -> only pixels have that
+        if data_schema.get_binsize() is None:
```
This is an explicit check for Pixels.

Mittmich replied:
Indeed, as discussed in the PR comment

dmitrymyl reviewed (changes requested):
Thanks for the quick changes @Mittmich! A couple of my comments are in the code and below:

1. Let's fix `OffsetAggregation` and the whole Pileup flow for Pixels only for now, maybe we will change it later.

2. Offset vs Distance: if we treat the query engine as bioframe/bedtools analogue, than we'd rather stick to the *distance* word, since it is the one used by these packages and *offset* is not found in their docs. I agree that *distance* is usually a non-negative metric, while here we have it otherwise, but *offset* might be even more confusing than that.

3. Windows constructed on the go from supplied regions &mdash; it works, let's leave it like that for now. I was thinking about outsourcing the creation of windows into a *complex analytical procedure*, i.e. we would have basic building blocks in the query engine and then a wrapper method/class that transforms Regions into windows. This still doesn't help with the fact that `OffsetTransformation` calculates distances to the midpoints of windows, which kinda leaks assumptions from an analytical procedure into the query engine, but this is okay for now.

Otherwise, besides some naming conventions, I would be happy to approve this PR to try the query engine in the wild and plan the next directions :)

Mittmich commented:
Thanks for the input, @dmitrymyl! I renamed offset to distance everywhere :) Let's merge and start using spoc!
